actions/file_actions.jsx

```diff
@@ -11,68 +10,101 @@ import {Client4} from 'mattermost-redux/client';
 
 import * as Utils from 'utils/utils';
 
-export function uploadFile(file, name, channelId, rootId, clientId) {
-    return (dispatch) => {
+export function uploadFile({file, name, type, rootId, channelId, clientId, onProgress, onSuccess, onError}) {
+    return (dispatch, getState) => {
         dispatch({type: FileTypes.UPLOAD_FILES_REQUEST});
 
-        return request.
-            post(Client4.getFilesRoute()).
-            set(Client4.getOptions({method: 'post'}).headers).
-
-            // The order here is important:
-            // keeping the channel_id/client_ids fields before the files contents
-            // allows the server to stream the uploads instead of loading them in memory.
-            field('channel_id', channelId).
-            field('client_ids', clientId).
-            attach('files', file, name).
-            accept('application/json');
-    };
-}
+        var xhr = new XMLHttpRequest();
 
-export function handleFileUploadEnd(file, name, channelId, rootId, clientId, {err, res}) {
-    return (dispatch, getState) => {
-        if (err) {
-            let e;
-            if (res && res.body && res.body.id) {
-                e = res.body;
-            } else if (err.status === 0 || !err.status) {
-                e = {message: Utils.localizeMessage('file_upload.generic_error', 'There was a problem uploading your files.')};
-            } else {
-                e = {message: Utils.localizeMessage('channel_loader.unknown_error', 'We received an unexpected status code from the server.') + ' (' + err.status + ')'};
+        xhr.open('POST', Client4.getFilesRoute(), true);
+
+        for (const header in Client4.getOptions({method: 'POST'}).headers) {
+            if (Client4.getOptions({method: 'POST'}).headers.hasOwnProperty(header)) {
```

Comment:
You can remove the `hasOwnProperty` check if you switch the outer loop from a for-in to a for-of loop

Reply:
removed for in loop but couldnt replace it with for of `[Type 'object' must have a 'Symbol.iterator' method that returns an iterator](https://stackoverflow.com/questions/59667515/type-object-must-have-a-symbol-iterator-method-that-returns-an-iterator)`
So adopted a safer object.keys method

Reply:
Ah, I missed that we were looping over an object there. You can still use `for (const header of Object.keys(headers)) {}` there as well or something fancier with `Object.entries`, but I'm 0/5 at that point